diff -r -u /usr/lib/python2.7/site-packages/pypungi/__init__.py /usr/lib/python2.7/site-packages/pypungi/__init__.py 
--- /usr/lib/python2.7/site-packages/pypungi/__init__.py	2013-11-01 01:18:51.000000000 +0000
+++ /usr/lib/python2.7/site-packages/pypungi/__init__.py	2014-01-03 17:16:37.290156293 +0000
@@ -774,7 +774,8 @@
 
         # get package objects according to the input list
         self.getPackageObjects()
-        self.createSourceHashes()
+        if self.is_sources:
+            self.createSourceHashes()
 
         pass_num = 0
         added = set()
@@ -799,8 +800,10 @@
                 for po in sorted(self.po_list):
                     added.update(self.get_package_deps(po))
 
-            added_srpms = self.add_srpms()
-            added.update(added_srpms)
+            if self.is_sources:
+                added_srpms = self.add_srpms()
+                added.update(added_srpms)
+
             if self.is_selfhosting:
                 for srpm_po in sorted(added_srpms):
                     added.update(self.get_package_deps(srpm_po))
@@ -816,7 +819,8 @@
             # add langpacks
             new = self.add_langpacks(self.po_list)
             self.langpack_packages.update(new)
-            self.langpack_packages.update([ self.sourcerpm_srpmpo_map[i.sourcerpm] for i in new ])
+            if self.is_sources:
+                self.langpack_packages.update([ self.sourcerpm_srpmpo_map[i.sourcerpm] for i in new ])
             added.update(new)
             if added:
                 continue
